### ISSUE TITLE
Dummy Pstream Library

### ISSUE THREAD

**Issue by jarlekramer** [mrklein/openfoam-os-x#15]
I have used your recipe before, with success, but after doing a clean install of OS X El Capitan I have problems. I get the error "Trying to use the dummy Stream library. This dummy library cannot be used in parallel mode" when running applications in parallel. Guessing it has something to do with open-mpi, but open-mpi is definitely installed via home-brew. Current version: open-mpi-1.10.1.

**Comment by mrklein** [comment]
The error usually means the order of folders in DYLD_LIBRARY_PATH is wrong. I.e. path to dummy Pstream libraries is before Pstream libraries linked with open-mpi.

Could you please post

1. OpenFOAM version (or even name of patch file you have used)
2. Contents of your DYLD_LIBRARY_PATH and LD_LIBRARY_PATH variables (though they should be identical). This could be done using

```sh
echo $DYLD_LIBRARY_PATH
echo $LD_LIBRARY_PATH
```

**Comment by jarlekramer** [comment]
OpenFoam version 3.0.0, and 2.4.0 tested

echo $DYLD_LIBRARY_PATH: /usr/local/opt/boost/lib:/usr/local/opt/cgal/lib:/Users/jarlekramer/OpenFOAM/OpenFOAM-3.0.0/platforms/darwin64ClangDPInt32Opt/lib/openmpi-system:/usr/local/Cellar/open-mpi/1.10.1/lib:/Users/jarlekramer/OpenFOAM/jarlekramer-3.0.0/platforms/darwin64ClangDPInt32Opt/lib:/Users/jarlekramer/OpenFOAM/site/3.0.0/platforms/darwin64ClangDPInt32Opt/lib:/Users/jarlekramer/OpenFOAM/OpenFOAM-3.0.0/platforms/darwin64ClangDPInt32Opt/lib:/Users/jarlekramer/OpenFOAM/OpenFOAM-3.0.0/platforms/darwin64ClangDPInt32Opt/lib/dummy

echo $LD_LIBRARY_PATH:
/usr/local/opt/boost/lib:/usr/local/opt/cgal/lib:/Users/jarlekramer/OpenFOAM/OpenFOAM-3.0.0/platforms/darwin64ClangDPInt32Opt/lib/openmpi-system:/usr/local/Cellar/open-mpi/1.10.1/lib:/Users/jarlekramer/OpenFOAM/jarlekramer-3.0.0/platforms/darwin64ClangDPInt32Opt/lib:/Users/jarlekramer/OpenFOAM/site/3.0.0/platforms/darwin64ClangDPInt32Opt/lib:/Users/jarlekramer/OpenFOAM/OpenFOAM-3.0.0/platforms/darwin64ClangDPInt32Opt/lib:/Users/jarlekramer/OpenFOAM/OpenFOAM-3.0.0/platforms/darwin64ClangDPInt32Opt/lib/dummy

**Comment by mrklein** [comment]
I have exactly (except user name) values of these variables. Just tested 3.0.0 and 2.4.0 cavity tutorial and it runs OK. So I need more details to reproduce your error.

Just to check obvious thing, do you have libPstream.dylib in $WM_PROJECT_DIR/platforms/darwin64ClangDPInt32Opt/lib/openmpi-system?

**Comment by jarlekramer** [comment]
yes

**Comment by jarlekramer** [comment]
Ah, sorry. Cavity runs fine for me as well. SnappyHexMesh does not though. Maybe the problem is related to snappy only...

**Comment by mrklein** [comment]
Could you provide exact steps to reproduce the error? Do you utilize Allrun script to execute the case (in this case it could be El Capitan's SIP, who clears DYLD_LIBRARY_PATH, so everything goes a little bit nuts)? I can try to continue guessing, but it would be easier for us both, if you post steps from start to the error.

**Comment by jarlekramer** [comment]
Everything is installed exactly as described in release+homebrew. When I test OpenFoam I use a python script that automatically generate a ship simulation case, including run scripts that are simple bash scripts. The script are as follows (the cellVolume command is a custom application that I use):

#!/bin/bash

surfaceFeatureExtract
blockMesh
decomposePar
mpirun -np 4 snappyHexMesh -overwrite -parallel
reconstructParMesh -constant
rm -fr processor*

COUNTER=0
while [ $COUNTER -lt 7 ]; do
	cellVolume
	topoSet -dict system/topoSetDict.$COUNTER
	refineMesh -overwrite
	let COUNTER=COUNTER+1
done

rm 0/cellVolume
renumberMesh -overwrite

**Comment by jarlekramer** [comment]
And since you mentioned scripts, I tested it manually as well. It is due to the script! When I run the commands manually, it works fine. Sorry for the confusion. Something in the script makes it so it does not work, but I guess the installation actually works fine then.

**Comment by jarlekramer** [comment]
The script works for executing the serial commands though, so somehow the script makes the parallel part stop working...

**Comment by jarlekramer** [comment]
Disabling SIP worked... Not sure what consequence that has though, so maybe it exist a better (safer?) solution? Anyhow, it seems it has nothing to do with OpenFOAM or your patches at least...

**Comment by mrklein** [comment]
Nuance with SIP is that it clears LD_LIBRARY_PATH (and DYLD_LIBRARY_PATH) for "protected" executables. So bash scripts live with empty LD_LIBRARY_PATH.

If you run executables by hand, they behave a little bit different.

I am in process of producing patches (and submitting certain part upstream) for foamJob and RunFunctions (runApplicaition and runParallel in particular) that account for OS X version 15.0.0.

**Comment by mrklein** [comment]
But on the other side I can not help with this:

```sh
#!/bin/bash

surfaceFeatureExtract
...
```

You execute surfaceFeatureExtract directly (my uncommitted fixes could help if you do `runApplication surfaceFeatureExtract`, runApplication function sets DYLD_LIBRARY_PATH variable). As a workaround just in the beginning of the script put

```sh
source /your/path/to/openfoam/etc/bashrc
```

to repopulate variables cleaned by SIP.

Guess, wiki page for El Capitan user could be added.

**Comment by mrklein** [comment]
Added wiki page on El Capitan's SIP. Right now I do not have solution except adding `source $WM_PROJECT_DIR/etc/bashrc` to the beginning of shell scripts (this workaround I tried to use in runApplication function, but I have not updated patches yet and now I think it would be overkill to source OpenFOAM's bashrc on every runApplication call).

**Comment by wyldckat** [comment]
Greetings to all!

I exchanged a few emails with a student who had some similar issues with Mac OS X El Capitan and the following should do the trick for you as well:

  1. Edit the file `$WM_PROJECT_DIR/bin/tools/RunFunctions`

  2. Find this line:
    ```
    ( mpirun -np $nProcs $APP_RUN -parallel "$@" < /dev/null > log.$APP_NAME 2>&1 )
    ```

  3. Replace it with this:
    ```
    ( mpirun -np $nProcs $(which foamExec) $APP_RUN -parallel "$@" < /dev/null > log.$APP_NAME 2>&1 )
    ```

The trick is `$(which foamExec)` which enforce using OpenFOAM's `foamExec` script, which will source the OpenFOAM environment and then calls the application with the environment up and running.

This in fact should be made the default in OpenFOAM, because this isn't the only situation where this can happen.

Best regards,
Bruno

**Comment by mrklein** [comment]
Dear Bruno,

Thank you for the comment, yet, unfortunately, I can not agree with the solution and especially with the part "...should be made the default in OpenFOAM...". Initially I planned to solve the problem similar way (though foamExec should also be added to runApplication function) but...

1. In fact this is necessary only on El Capitan with activated SIP (on other platforms bash scripts enjoy unmodified (DY*)LD_LIBRARY_PATH variables).
2. This would be overkill if person have several runApplication or runParallel calls in one script.

So I have decided to add the following to `$WM_PROJECT_DIR/bin/tools/RunFunctions`:

```sh
isSIPEnabled()
{
    [ which -s csrutil ] && {
    [ $(csrutil status | awk '{print $5}' | tr -d ".") = "enabled" ] \
        && return 0
    }
    return 1
}

# Workaround for cleared DYLD_LIBRARY_PATH
[ "$(uname -s)" = "Darwin" ] && [ isSIPEnabled ] && . $WM_PROJECT_DIR/etc/bashrc
```

Basically this is what foamExec does unconditionally:

```sh
sourceRc
exec "$@"
```

I have added modified RunFunctions file to repository: https://github.com/mrklein/openfoam-os-x/blob/master/files/RunFunctions.

**Comment by wyldckat** [comment]
Dear Alexey,

Sorry, I forgot to answer to your comment a few weeks ago.
Thumbs up on the implementation you've done! And I do agree with you that always calling `foamExec` is overkill.

Best regards,
Bruno
